src/psyclone/psyGen.py

```diff
@@ -5832,12 +5861,127 @@ def _number_handler(self, node, parent):
+        self._is_arg = argument
+        self._access = SymbolAccess.UNKNOWN
+        if access:
+            self.access = access
+
+    @property
+    def is_argument(self):
+        '''
+        :returns: Whether or not this symbol is passed as a routine argument.
+        :rtype: bool
+        '''
+        return self._is_arg
+
+    @property
+    def access(self):
+        '''
+        :returns: the access-type for this symbol.
+        :rtype: :py:class:`psyclone.psyGen.SymbolAccess`
+        '''
+        return self._access
+
+    @access.setter
+    def access(self, value):
+        '''
+        Setter for the access type of this symbol.
+
+        :param value: the new access type.
+        :type value: :py:class:`psyclon.psyGen.SymbolAccess`
+
+        :raises TypeError: if the supplied value is not of the correct type.
+        '''
+        if not isinstance(value, SymbolAccess):
+            raise TypeError("SymbolInterface.access must be a 'SymbolAccess' "
+                            "but got '{0}'.".format(type(value)))
+        self._access = value
+
+    def __str__(self):
+        return "Argument={0}".format(self.is_argument)
+
+
+class FortranInterface(SymbolInterface):
+    '''
+    Describes the interface to a Fortran Symbol. If no arguments are passed
+    to the constructor then the interface represents a routine argument
+    with the default INTENT of INOUT.
+
+    :param access: the manner in which the Symbol is accessed in the \
+                   associated code section. If None is supplied then the \
+                   access is SymbolAccess.UNKNOWN.
+    :type access: :py:class:`psyclone.psyGen.SymbolAccess` or None.
+    :param str module_use: the name of the Fortran module from which the \
+                           symbol is imported (or None if this symbol is a \
+                           routine argument).
+    '''
+    def __init__(self, access=None, module_use=None):
+        # By default a FortranInterface represents a routine argument
+        argument = False
+        if not module_use:
+            argument = True
+        super(FortranInterface, self).__init__(argument=argument,
+                                               access=access)
+        if self.is_argument and not access:
+            # In Fortran a routine argument without an explicit INTENT
+            # attribute has intent INOUT
+            self._access = SymbolAccess.READWRITE
+        self._module_name = module_use
+
+    def __str__(self):
+        if self.module_name:
+            return "FortranModule({0})".format(self.module_name)
+        else:
+            return super(FortranInterface, self).__str__()
+
+    @property
+    def module_name(self):
+        '''
+        :returns: the name of the Fortran module from which the symbol is \
+                  imported or None if it is not a module variable.
+        :rtype: str or None
+        '''
+        return self._module_name
```

Comment:
With the current `Interface` and `FortranInterface` is not very clear when you read the user code that we are creating arguments or globals. At the moment we have:
``` 
Symbol('a', 'real', [], interface=FortranInterface(access=...))
Symbol('b', 'real', [], interface=FortranInterface(access=..., module_use=...))
```
So we just know that `a` is an Argument because it is the default of FortranInterface and `b` is a global because it has the `module_use` argument.  I think a more readable alternative would be to have `Interface` as an Abstract base class with the `is_argument` and `is_global` methods defined, and then implement Argument and FortranGlobal as subclasses. Also I believe the access parameter is not needed for FortranGlobals as they are always READWRITE, right? An then optionally this can also be a nested (just has meaning for Symbol). With all that we could get some user interface like this:
``` 
Symbol('a', 'real', [], interface=Symbol.Argument(access=...))
Symbol('b', 'real', [], interface=Symbol.FortranGlobal(module_use=...))
```
What do you think?

Reply:
I think that's a good idea and have re-factored things accordingly.